### PR TITLE
Allocate empty group on heap only for `std::allocator`

### PR DESCRIPTION
**Pull request by greg7mdp**
Otherwise leave `ctrl_` as a nullptr.

### CODE REVIEW

**Comment by AgostonSzepessy** [comment]
Thanks for implementing this! Do you know if it would result in a perf regression if we didn't have `ctrl_`? The other thing we could do is user opt-in to store the `ctrl_` on the `raw_hash_set` if there's a big perf gap.

**Comment by greg7mdp** [comment]
> Thanks for implementing this! Do you know if it would result in a perf regression if we didn't have `ctrl_`? The other thing we could do is user opt-in to store the `ctrl_` on the `raw_hash_set` if there's a big perf gap.

Regarding the perf regression, I very much doubt you could see any difference in real world use. This is why I prefer optimizing for correctness and overall good behavior (for example with poor hash functions) rather than for micro benchmarks.

I don't think storing the `ctrl_` on the `raw_hash_set` is a good idea. First it probably would make no difference, and also it would do an allocation and use memory for empty hash maps, which I think is a bigger issue.

**Comment by greg7mdp** [comment]
Thanks again for bringing up this issue @AgostonSzepessy !

**Comment by AgostonSzepessy** [comment]
Thank you for fixing it so quickly!

File: parallel_hashmap/phmap.h [hunk 1]
```diff
@@ -329,11 +329,16 @@ static_assert(kDeleted == -2,
...
-      kEmpty,    kEmpty, kEmpty, kEmpty, kEmpty, kEmpty, kEmpty, kEmpty};
-  return const_cast<ctrl_t*>(empty_group);
+  PHMAP_IF_CONSTEXPR (std_alloc_t::value) {
+      alignas(16) static constexpr ctrl_t empty_group[] = {
+          kSentinel, kEmpty, kEmpty, kEmpty, kEmpty, kEmpty, kEmpty, kEmpty,
+          kEmpty,    kEmpty, kEmpty, kEmpty, kEmpty, kEmpty, kEmpty, kEmpty};
+
+      return const_cast<ctrl_t*>(empty_group);
+  }
+  return nullptr;
```
Comment (vm2mv) [inline parallel_hashmap/phmap.h:341]: Warning in VS 2022:
`phmap.h(341): warning C4702: unreachable code`

Comment (greg7mdp) [reply parallel_hashmap/phmap.h:341]: Thanks. Fixed in https://github.com/greg7mdp/parallel-hashmap/commit/2d0b499273ea02f7211ae2af705418c9509691fb.

**Pull request merged**